programs/dynamic-bonding-curve/src/instructions/migration/meteora_damm/meteora_damm_metadata_state.rs

```diff
@@ -31,12 +51,248 @@ pub struct MeteoraDammMigrationMetadata {
+
+        self.inner.set_creator_claim_status();
+        self.inner.set_partner_claim_status();
+
+        Ok(total_lp_to_claim)
+    }
+
+    fn validate_and_claim_as_creator(&mut self) -> Result<u64> {
+        self.inner.common_validate_creator_claim_lp_action()?;
+
+        let total_lp_to_claim = self.inner.creator_lp;
+        require!(total_lp_to_claim != 0, PoolError::NotPermitToDoThisAction);
+
+        self.inner.set_creator_claim_status();
+
+        Ok(total_lp_to_claim)
+    }
+
+    fn validate_and_claim_as_partner(&mut self) -> Result<u64> {
+        self.inner.common_validate_partner_claim_lp_action()?;
+
+        let total_lp_to_claim = self.inner.partner_lp;
+        require!(total_lp_to_claim != 0, PoolError::NotPermitToDoThisAction);
+
+        self.inner.set_partner_claim_status();
+
+        Ok(total_lp_to_claim)
+    }
+}
+
+struct MigrationMetadataV2<'a> {
+    inner: &'a mut MeteoraDammMigrationMetadata,
+}
+
+impl<'a> MigrationMetadataImplementation<'a> for MigrationMetadataV2<'a> {
+    fn validate_and_lock_as_creator(&mut self, accounts: DammAccounts<'_, '_>) -> Result<u64> {
+        self.inner.common_validate_creator_lock_lp_action()?;
+        let actual_lock_amount = exclude_fee_lp_amount(self.inner.creator_locked_lp, accounts)?;
+        self.inner.actual_creator_locked_lp = actual_lock_amount;
+        self.inner.set_creator_lock_status();
+        Ok(actual_lock_amount)
+    }
+
+    fn validate_and_lock_as_partner(&mut self, accounts: DammAccounts<'_, '_>) -> Result<u64> {
+        self.inner.common_validate_partner_lock_lp_action()?;
+        let actual_lock_amount = exclude_fee_lp_amount(self.inner.partner_locked_lp, accounts)?;
+        self.inner.actual_partner_locked_lp = actual_lock_amount;
+        self.inner.set_partner_lock_status();
+        Ok(actual_lock_amount)
+    }
+
+    fn validate_and_lock_as_self_partnered_creator(
+        &mut self,
+        accounts: DammAccounts<'_, '_>,
+    ) -> Result<u64> {
+        self.inner
+            .common_validate_self_partnered_creator_lock_lp_action()?;
+
+        let total_lp_to_lock = self
+            .inner
+            .creator_locked_lp
+            .safe_add(self.inner.partner_locked_lp)?;
+
+        let actual_lock_amount = exclude_fee_lp_amount(total_lp_to_lock, accounts)?;
+
+        self.inner.actual_partner_locked_lp = actual_lock_amount
+            .safe_mul(self.inner.partner_locked_lp)?
+            .safe_div(total_lp_to_lock)?;
+
+        self.inner.actual_creator_locked_lp =
+            actual_lock_amount.safe_sub(self.inner.actual_partner_locked_lp)?;
+
+        self.inner.set_creator_lock_status();
+        self.inner.set_partner_lock_status();
+
+        Ok(actual_lock_amount)
+    }
+
+    fn validate_and_claim_as_self_partnered_creator(&mut self) -> Result<u64> {
+        self.inner
+            .common_validate_self_partnered_creator_claim_lp_action()?;
+
+        // Must lock first to know actual locked lp amount
```

Comment:
What happen if percentage of locked lp is zero? they actually they dont need to lock, just claim directly

Reply:
Fixed